src/Surfnet/StepupMiddleware/CommandHandlingBundle/SensitiveData/EventSourcing/SensitiveDataMessageStream.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/**
+ * Copyright 2014 SURFnet bv
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\EventSourcing;
+
+use Broadway\Domain\DomainEventStreamInterface;
+use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\Exception\SensitiveDataApplicationException;
+use Surfnet\StepupMiddleware\CommandHandlingBundle\SensitiveData\Forgettable;
+
+final class SensitiveDataMessageStream
+{
+    /**
+     * @var array
+     */
+    private $messages;
+
+    /**
+     * @param SensitiveDataMessage[] $messages
+     */
+    public function __construct(array $messages)
+    {
+        $this->messages = $messages;
+    }
+
+    public function applyToDomainEventStream(DomainEventStreamInterface $domainEventStream)
+    {
+        reset($this->messages);
+
+        /** @var DomainMessage $domainMessage */
+        foreach ($domainEventStream as $domainMessage) {
+            /** @var SensitiveDataMessage|false $sensitiveDataMessage */
+            $sensitiveDataMessage = current($this->messages);
+
+            if (!$sensitiveDataMessage || $sensitiveDataMessage->getPlayhead() != $domainMessage->getPlayhead()) {
+                continue;
+            }
+
+            $event = $domainMessage->getPayload();
+
+            if (!$event instanceof Forgettable) {
+                throw new SensitiveDataApplicationException(
+                    'Encountered sensitive data for event which does not support it'
+                );
+            }
+
+            $event->setSensitiveData($sensitiveDataMessage->getSensitiveData());
+
+            next($this->messages);
+        }
```

Comment:
Might deserve a check for the inverse variants too: 
- an event that implements `Forgettable` but has no corresponding SensitiveData as well as SensitiveData objects that have no events and are thus not applied (for instance at the end of the SensitiveDataMessageStream)